pretyping/coercion.ml

```diff
@@ -522,52 +522,97 @@ let reapply_coercions_body sigma trace body =
+      Pp.(str "Hook already registered: \"" ++ str name ++ str "\".");
+  all_hooks := CString.Map.add name h !all_hooks
+
+let active_hooks = Summary.ref ~name:"coercion_hooks" ([] : string list)
+
+let deactivate_hook ~name =
+  active_hooks := List.filter (fun s -> not (String.equal s name)) !active_hooks
+
+let activate_hook ~name =
+  assert (CString.Map.mem name !all_hooks);
+  deactivate_hook ~name;
+  active_hooks := name :: !active_hooks
+
+let active_hooks () =
+  List.map (fun name -> CString.Map.get name !all_hooks) !active_hooks
+
+let default_flags_of env =
+  default_flags_of TransparentState.full
+
 (* Try to coerce to a funclass; raise NoCoercion if not possible *)
 let inh_app_fun_core ~program_mode ?(use_coercions=true) env sigma body typ =
   match unify_product env sigma typ with
-  | Inl sigma -> sigma, body, typ, IdCoe
+  | Inl sigma -> sigma, body, Some typ, IdCoe
   | Inr t ->
     try
       if not use_coercions then raise NoCoercion;
       let p = lookup_path_to_fun_from env sigma typ in
       let body = force_app_body body in
       let sigma, body, typ, trace = apply_coercion env sigma p body typ in
-      sigma, start_app_body sigma body, typ, trace
+      sigma, start_app_body sigma body, Some typ, trace
     with (Not_found | NoCoercion) as exn ->
       let _, info = Exninfo.capture exn in
       if program_mode then
         try
           let sigma, (coercef, t, trace) = mu env sigma t in
           let j = {uj_val=force_app_body body; uj_type = typ} in
           let sigma, uj_val = app_opt env sigma coercef j.uj_val in
-          (sigma, start_app_body sigma uj_val, t, trace)
+          (sigma, start_app_body sigma uj_val, Some t, trace)
         with NoSubtacCoercion | NoCoercion ->
-          (sigma,body,typ,IdCoe)
+          (sigma,body,Some typ,IdCoe)
       else Exninfo.iraise (NoCoercion,info)
 
 (* Try to coerce to a funclass; returns [j] if no coercion is applicable *)
-let inh_app_fun ~program_mode ~resolve_tc ?use_coercions env sigma body typ =
-  try inh_app_fun_core ~program_mode ?use_coercions env sigma body typ
+let inh_app_fun ~program_mode ~resolve_tc ?use_coercions env sigma ?(flags=default_flags_of env) body typ =
+  try
+    try inh_app_fun_core ~program_mode ?use_coercions env sigma body typ
+    with
+    | NoCoercion when resolve_tc
+      && (get_use_typeclasses_for_conversion ()) ->
+        inh_app_fun_core ~program_mode ?use_coercions env (saturate_evd env sigma) body typ
   with
-  | NoCoercion when not resolve_tc
-    || not (get_use_typeclasses_for_conversion ()) -> (sigma, body, typ, IdCoe)
-  | NoCoercion ->
-    try inh_app_fun_core ~program_mode ?use_coercions env (saturate_evd env sigma) body typ
-    with NoCoercion -> (sigma, body, typ, IdCoe)
+  | NoCoercion -> let hook_res =
+      List.fold_left
+        (fun r h ->
+          if r <> None then r else h env sigma ~flags (force_app_body body) ~inferred:typ ~expected:Product)
+        None (active_hooks ()) in
```

Comment:
Instead of fold_left you can use find_map

Reply:
also you should factor this and the other hook calls as `apply_hooks env sigma ~flags body ~inferred ~expected = CList.find_map (fun h -> h env sigma ~flags body ~inferred ~expected) (active_hooks())`, then this code becomes just 
```suggestion
      apply_hooks env sigma ~flags (force_app_body body) ~inferred ~expected:Product
```